x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelCacheMetadataService.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.persistence;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor.TaskContext;
+import org.elasticsearch.cluster.ClusterStateTaskListener;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.cluster.service.MasterServiceTaskQueue;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xpack.core.XPackPlugin;
+import org.elasticsearch.xpack.core.ml.action.FlushTrainedModelCacheAction;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelCacheMetadata;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+public class TrainedModelCacheMetadataService implements ClusterStateListener {
+    private static final Logger LOGGER = LogManager.getLogger(TrainedModelCacheMetadataService.class);
+    static final String TASK_QUEUE_NAME = "trained-models-cache-metadata-management";
+    private final MasterServiceTaskQueue<CacheMetadataUpdateTask> metadataUpdateTaskQueue;
+    private final Client client;
+    private volatile boolean isMasterNode = false;
+
+    public TrainedModelCacheMetadataService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, ML_ORIGIN);
+        CacheMetadataUpdateTaskExecutor metadataUpdateTaskExecutor = new CacheMetadataUpdateTaskExecutor();
+        this.metadataUpdateTaskQueue = clusterService.createTaskQueue(TASK_QUEUE_NAME, Priority.IMMEDIATE, metadataUpdateTaskExecutor);
+        clusterService.addListener(this);
+    }
+
+    public void refreshCacheVersion(ActionListener<AcknowledgedResponse> listener) {
```

Comment:
```suggestion
    public void invalidateCache(ActionListener<AcknowledgedResponse> listener) {
```

Or `updateCacheVersion()` makes more sense as a function name to me

Reply:
Renamed to updateCacheVersion